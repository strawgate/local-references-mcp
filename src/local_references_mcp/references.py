from datetime import datetime
from pathlib import Path

from fastmcp.contrib.mcp_mixin import MCPMixin, mcp_tool
from pydantic import BaseModel, Field


class ReferenceEntry(BaseModel):
    name: str = Field(description="The name of the reference entry")
    path: Path = Field(description="The path to the reference entry file")

    def get_length(self) -> int:
        with self.path.open("r") as f:
            return len(f.read())

    def get_content(self, head: int | None = None) -> str:
        with self.path.open("r") as f:
            return f.read()[:head]

    def get_preview(self, preview_size: int) -> str:
        content = self.get_content(preview_size)

        # Get everything before the first double newline
        content = content.split("\n\n")[0] if "\n\n" in content else content

        # Truncate to the preview size
        content = content[:preview_size]

        # If the content is longer than the preview size, add an ellipsis
        if self.get_length() > preview_size:
            content = "Preview: " + content + "..."

        return content

    def render(self, reference_type: str, preview: bool, preview_size: int) -> str:
        contents = [
            f"### Type: `{reference_type}`, Name: `{self.name}`\n",
            self.get_content() if not preview else self.get_preview(preview_size),
            "\n",
        ]
        return "".join(contents)


class Reference(BaseModel):
    type: str = Field(description="The type of reference")
    path: Path = Field(description="The path to the reference file")

    @property
    def description(self) -> str:
        readme = self.path / "README.md"
        if readme.exists():
            return readme.read_text()
        return ""

    @classmethod
    def from_reference_string(cls, reference_string: str) -> "Reference":
        if ":" in reference_string:
            reference_type, path = reference_string.split(":")
            reference_path = Path(path)
        else:
            reference_path = Path(reference_string)
            reference_type = reference_path.stem

        return cls(type=reference_type, path=reference_path)

    def get_entries(self) -> list[ReferenceEntry]:
        # Get the markdown files in the directory
        return [ReferenceEntry(name=file.stem, path=file) for file in self.path.glob("*.md") if file.stem != "README"]

    def get_entry_by_name(self, name: str) -> ReferenceEntry:
        entries = self.get_entries()
        for entry in entries:
            if entry.name == name:
                return entry
        msg = f"No entry found with name {name}"
        raise ValueError(msg)

    def has_entry(self, name: str) -> bool:
        try:
            self.get_entry_by_name(name)
        except ValueError:
            return False

        return True

    def render(self, preview: bool, preview_size: int) -> str:
        contents = [
            f"## References for: `{self.type}`\n",
            self.description,
            "\n\n",
            "\n".join([entry.render(self.type, preview, preview_size) for entry in self.get_entries()]),
            "\n",
        ]
        return "".join(contents)


class ReferenceManager(MCPMixin):
    references: list[Reference]
    last_refreshed: datetime
    preview_size: int

    def __init__(self, references: list[Reference], preview_size: int):
        self.preview_size = preview_size
        self.references = references

    @classmethod
    def from_reference_strings(cls, reference_strings: list[str], preview_size: int) -> "ReferenceManager":
        return cls([Reference.from_reference_string(reference) for reference in reference_strings], preview_size)

    @mcp_tool()
    def preview_references(self) -> str:
        """Preview all references

        Returns:
            Formatted text describing the types of references and their available entries

        Example:
        >>> preview_references()
        # Local References

        ## Reference Type: Best Practices
        This is a description of the README reference type.

        ### Thing 1
        This is a preview of the best practice for thing 1.

        ### Thing 2
        This is a preview of the best practice for thing 2.
        """
        content = [
            "# Local References\n",
            "Below is a list of available reference types. We will show each type, and a preview of its entries.",
            f"Note: previews are truncated to {self.preview_size} characters.\n\n",
            "\n".join(
                [
                    reference.render(
                        preview=True,
                        preview_size=self.preview_size,
                    )
                    for reference in self.references
                ]
            ),
            "\n",
        ]
        return ("".join(content)).strip()

    @mcp_tool()
    def get_reference(self, reference_type: str, reference_name: str) -> str:
        """Get a reference by type and name

        Args:
            reference_type: The type of reference to get
            reference_name: The name of the reference to get

        Returns:
            The content of the reference

        Example:
        >>> get_reference("Best Practices", "Thing 1")
        This is a best practice for thing 1. It goes on much longer than the preview
        """
        for reference in self.references:
            if reference.type == reference_type:
                return reference.get_entry_by_name(reference_name).get_content()

        msg = f"No reference found with type {reference_type} and name {reference_name}"
        raise ValueError(msg)
