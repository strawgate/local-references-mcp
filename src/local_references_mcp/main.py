import asyncio

import asyncclick as click
from fastmcp import FastMCP

from local_references_mcp.references import ReferenceManager


@click.command()
@click.option("--reference", type=str, required=True, multiple=True, help="The reference to use. name:path")
@click.option("--preview-size", type=int, default=1000, help="The number of characters to preview from each entry")
async def cli(reference: list[str], preview_size: int):
    mcp = FastMCP(name="Local References")

    reference_manager: ReferenceManager = ReferenceManager.from_reference_strings(reference, preview_size)

    reference_manager.register_tools(mcp_server=mcp)

    await mcp.run_async()


def run_mcp():
    asyncio.run(cli())


if __name__ == "__main__":
    run_mcp()
