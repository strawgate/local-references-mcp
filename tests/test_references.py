import tempfile
from pathlib import Path

import pytest

from local_references_mcp.references import Reference, ReferenceEntry, ReferenceManager


@pytest.fixture
def temp_directory():
    with tempfile.TemporaryDirectory() as temp_dir:
        yield Path(temp_dir)


def test_reference_entry_get_content(temp_directory: Path):
    temp_file = temp_directory / "Dashboard.md"
    temp_file.write_text("This is a test reference.")
    entry = ReferenceEntry(name="Dashboard", path=temp_file)
    content = entry.get_content()
    assert "This is a test reference." in content


def test_reference_entry_get_preview(temp_directory: Path):
    temp_file = temp_directory / "Dashboard.md"
    temp_file.write_text("This is a test reference.\n\nEntry1 content.")
    entry = ReferenceEntry(name="Dashboard", path=temp_file)
    preview = entry.get_preview(preview_size=1000)
    assert "Entry1 content." not in preview  # Should only get up to first double newline


def test_reference_from_reference_string(temp_directory: Path):
    temp_file = temp_directory / "Dashboard.md"
    ref_str = f"TestRef:{temp_file}"
    ref = Reference.from_reference_string(ref_str)
    assert ref.type == "TestRef"
    assert str(ref.path) == str(temp_file)


def test_reference_get_entry_by_name(temp_directory: Path):
    # Create a directory with a README.md file
    temp_file = temp_directory / "README.md"
    temp_file.write_text("This is the description of this reference type.")

    # Create a directory with a How To Make a Sandwich.md file
    how_to_file = temp_directory / "How To Make a Sandwich.md"
    how_to_file.write_text("This is the content of the how-to reference.")

    ref = Reference(type="How To", path=temp_directory)

    # Get the entry by name
    entry = ref.get_entry_by_name("How To Make a Sandwich")
    assert entry.path == how_to_file


def test_reference_get_description(temp_directory: Path):
    # Create a directory with a README.md file
    temp_file = temp_directory / "README.md"
    temp_file.write_text("This is the description of this reference type.")

    ref = Reference(type="How To", path=temp_directory)
    assert ref.description == "This is the description of this reference type."


def test_reference_manager_and_preview(temp_directory: Path):
    temp_file = temp_directory / "README.md"
    temp_file.write_text("This is the description of this reference type.")

    temp_entry = temp_directory / "Dashboard.md"
    temp_entry.write_text("This is a test reference.")

    ref_str = f"Cool Reference Type:{temp_directory}"
    mgr = ReferenceManager.from_reference_strings([ref_str], preview_size=1000)
    preview = mgr.preview_references()
    assert "# Local References" in preview
    assert "description of this reference type" in preview
    assert "Dashboard" in preview
    assert "This is a test reference." in preview


def test_reference_manager_get_reference(temp_directory: Path):
    # Create the readme
    temp_file = temp_directory / "README.md"
    temp_file.write_text("This is the description of this reference type.")

    # Create an entry
    entry_file = temp_directory / "Making a Sandwich.md"
    entry_file.write_text("This is a test reference.")

    ref_str = f"Best Practices:{temp_directory}"
    mgr = ReferenceManager.from_reference_strings([ref_str], preview_size=1000)
    content = mgr.get_reference("Best Practices", "Making a Sandwich")
    assert "This is a test reference." in content


def test_reference_structure_snapshot(tmp_path, snapshot):
    # Create a reference structure with two files

    folder1 = tmp_path / "Best Practices"
    folder1.mkdir(parents=True, exist_ok=True)
    file1 = folder1 / "README.md"
    file1.write_text("This is the best practice reference type.")
    file2 = folder1 / "Making a Sandwich.md"
    file2.write_text("This is the best practice reference.\n\nEntry1 content.")
    folder2 = tmp_path / "How To"
    folder2.mkdir(parents=True, exist_ok=True)
    file3 = folder2 / "README.md"
    file3.write_text("This is the how-to reference type.")
    file4 = folder2 / "How To Make a Sandwich.md"
    file4.write_text("This is the how-to reference.\n\nHowTo entry content.")
    file5 = folder2 / "How To Make a Sandwich.md"
    file5.write_text(
        "Of resolve to gravity thought my prepare chamber so. Unsatiable entreaties collecting may sympathize nay interested instrument. If continue building numerous of at relation in margaret. Lasted engage roused mother an am at. Other early while if by do to. Missed living excuse as be. Cause heard fat above first shall for. My smiling to he removal weather on anxious. "  # noqa: E501
        * 3
    )

    ref_strs = [f"Best Practices:{folder1}", f"How To:{folder2}"]
    mgr = ReferenceManager.from_reference_strings(ref_strs, preview_size=1000)

    # Snapshot the preview
    assert snapshot == mgr.preview_references()

    # Snapshot the full rendering for one reference
    assert snapshot == mgr.get_reference("Best Practices", "Making a Sandwich")
