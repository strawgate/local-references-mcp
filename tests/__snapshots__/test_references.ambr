# serializer version: 1
# name: test_reference_structure_snapshot
  '''
  # Local References
  Below is a list of available reference types. We will show each type, and a preview of its entries.Note: previews are truncated to 1000 characters.
  
  ## References for: `Best Practices`
  This is the best practice reference type.
  
  ### Type: `Best Practices`, Name: `Making a Sandwich`
  This is the best practice reference.
  
  
  ## References for: `How To`
  This is the how-to reference type.
  
  ### Type: `How To`, Name: `How To Make a Sandwich`
  Preview: Of resolve to gravity thought my prepare chamber so. Unsatiable entreaties collecting may sympathize nay interested instrument. If continue building numerous of at relation in margaret. Lasted engage roused mother an am at. Other early while if by do to. Missed living excuse as be. Cause heard fat above first shall for. My smiling to he removal weather on anxious. Of resolve to gravity thought my prepare chamber so. Unsatiable entreaties collecting may sympathize nay interested instrument. If continue building numerous of at relation in margaret. Lasted engage roused mother an am at. Other early while if by do to. Missed living excuse as be. Cause heard fat above first shall for. My smiling to he removal weather on anxious. Of resolve to gravity thought my prepare chamber so. Unsatiable entreaties collecting may sympathize nay interested instrument. If continue building numerous of at relation in margaret. Lasted engage roused mother an am at. Other early while if by do to. Missed livi...
  '''
# ---
# name: test_reference_structure_snapshot.1
  '''
  This is the best practice reference.
  
  Entry1 content.
  '''
# ---
